File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> permuSet = new ArrayList<>();
        helper(nums,new ArrayList<>(),permuSet);
        return permuSet;
    }

    public void helper(int[] nums, List<Integer> temp, List<List<Integer>> permuSet) {
        if (temp.size() == nums.length) {
             permuSet.add(new ArrayList<>(temp));
        } else {
            for (int i = 0; i < nums.length; i++) {
                if(temp.contains(nums[i])){
                    continue;
                }
                temp.add(nums[i]);
                helper(nums,temp,permuSet);
                temp.remove(temp.size()-1);
            }
        }
    }
}

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        #Solution 1
        # ans = []
        # freq = [False] * len(nums)
        # ds = []
        # def helper(nums,ds,ans,freq):
        #     if len(ds) == len(nums):
        #         ans.append(ds[:])
        #         return 
        #     for i in range(len(nums)):
        #         if not freq[i]:
        #             freq[i] = True
        #             ds.append(nums[i])
        #             helper(nums,ds,ans,freq)
        #             ds.pop()
        #             freq[i] = False
        # helper(nums,ds,ans,freq)
        # return ans

        #Solution 2
        ans = []
        def helper(index,nums,ans):
            if index == len(nums):
                # for i in nums:
                #     ans.append([i])
                ans.append(nums[:])
                return 
            for i in range(index,len(nums)):
                swap(i,index,nums)
                helper(index+1,nums,ans)
                swap(i,index,nums)
        def swap(i,j,nums):
            print("adsfg")
            nums[i],nums[j] = nums[j],nums[i]
        helper(0,nums,ans)
        return ans
        
        
        